Add "Only in edit controls" mode for typing echo

cary-rowen opened the pull request:
### Link to issue number:
Fixes #16848, related #10331, #3027

### Summary of the issue:
Currently NVDA can only toggle typing echo (characters and words) on or off globally. Users want more granular control to only have typing feedback in edit controls, while keeping it off in other contexts like listss or non-edit areas.
### Description of user facing changes
- Added a new option "Only in edit controls" for both "Speak typed characters" and "Speak typed words" settings in Keyboard Settings
- Instead of checkboxes, these are now combo boxes with three options:
  - Off: No typing echo
  - Only in edit controls: Only echo text typed in edit fields
  - Always: Echo all typed text
- By default, "Speak typed characters" is now set to "Only in edit controls".
- Updated relevant documentation in the user guide

### Description of development approach
The implementation:
1. Added a TypingEcho enum in configFlags.py with values:
   - OFF (0)
   - EDIT_CONTROLS (1)
   - ALWAYS (2) 
2. Changed keyboard typing echo configuration from boolean to integer values
3. Updated speech.py, behaviors.py and inputComposition.py to use the new enum
4. Modified settings dialog to use combo box instead of checkbox
5. Updated documentation

### Testing strategy:
Tested the following scenarios:
1. Basic functionality:
- Open Notepad (edit control)
  - Set to "Only in edit controls"
  - Type text - should be announced
  - Verify both character and word echo settings
- Explorer file lists(non-edit control)
  - Type text - should not be announced
- Test all three modes (Off, On, Only in edit controls)
2. Different contexts:
- Web browser input fields
- Rich text editors
- Read-only text areas
- Terminal windows

### Known issues with pull request:
None identified.

### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [X] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [X] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [X] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [ ] API is compatible with existing add-ons.
- [ ] Security precautions taken.

<!-- Please keep the following -->
@coderabbitai summary

cary-rowen committed:
Add "Only in edit controls" mode for typing echo

pre-commit-ci[bot] committed:
Pre-commit auto-fix

Adriani90 commented:
Great work on this, thank you.

Adriani90 commented:
I think this needs a changelog entry as well.

cary-rowen committed:
Updated changes

cary-rowen committed:
Merge branch 'TypedWords' of github.com:cary-rowen/nvda into TypedWords

cary-rowen committed:
Merge branch 'master' into TypedWords

cary-rowen committed:
Add #noqa: E722 to suppress lint warnings

CyrilleB79 commented:
Very nice addition @cary-rowen! Thanks.

I have some questions:
1. Shouldn't there be a config upgrade? I.e. a value in the old config should remain in the new one.
2. I think that "Only in edit controls" should become the default behaviour; I guess that most people would be happy and moreover, it's a security fix as described in #10331.
3. If NV Access reviewers (@SaschaCowley or @seanbudd) do not accept the new behaviour to be the default for now, should we use a feature flag to be able to change the default in the future?
4. With the new option, can typing be heard when typing in an edit field but when text is not actually written, e.g.:
   * In cmd, when initiating an ssh connection, typing the password outputs no visible character
   * In a read-only edit field, e.g. the synth field of NVDA speech settings panel.

nvdaes reviewed (commented):
@cary-rowen , thanks for this.
I think that you may add a helper function in globalCommands, similar to the `toggleBooleanValue` function, for config flags, and use it for toggling typed characters and words.
For reference, see `toggleBooleanValue` in `globalCommands.py`, introduced in PR #16994

cary-rowen committed:
Update configuration schema to version 14 and add upgrade step for boolean to integer conversion.

pre-commit-ci[bot] committed:
Pre-commit auto-fix

cary-rowen commented:
Hi @CyrilleB79 
Thank you for your review! I'd like to clarify your last point about special input scenarios. Are you asking whether the new "Only in edit controls" option works as expected in these cases:
1. When typing passwords in cmd during SSH connection:
  - Should typed characters be announced in this case?
2. In read-only edit fields (like the synthesizer field in NVDA speech settings):
  - Though it looks like an edit field, it's read-only
  - I've tested that typed characters are not announced here with the new option, as no text can be entered
  - Is this the expected behavior?

Could you please confirm if these behaviors are what you're asking about, and what the expected behavior should be in these cases?

Many thanks

cary-rowen commented:
Regarding default settings, I'd also love to hear suggestions from NV Access, I'd personally like to see new options provided as defaults for the same reasons as @CyrilleB79.

@nvdaes 
Thanks for your suggestion, I will implement a helper method to simplify the code.

CyrilleB79 commented:
@cary-rowen IMO, the expected behaviour with the new option in ssh password case and in read-only edit fields such as synthesizer is that NVDA should not speak anything since nothing is actually visible on the screen.

amirmahdifard commented:
thaaaaanks! should I really beleave that from now on we can get rid of the t. This pc 10 of 18, or space. playback stopped, and space. Playback resumed, with out needing to install the best ever addon made by @cary-rowen that addon was the best, because I alwayes liked and needed to keep typed characters and words, but I hated how they are announced everywhere, I was thinking, how stupid, the option it self is called speeck ('"typed"') characters and words, but why whe were hearing it everywhere? whe are not typing anything! whe are using our keyboard to use and explor our computer! but from now on, at least we don't have to use an addon to solv this really annoying this. And, really thank you @cary-rowen for resolving such problem for us until now. befor your addon releace on the stor, I was using an old and less buggy vertion of an unknown addon called typing settings to solv this for my self. greate feature!

Adriani90 commented:
@cary-rowen speaking typed characters outside of edit fields does not make sense anyway, because they are not typed, but only pressed. For that we have the input help feature already if people need to learn the keyboard layout outside of edit fields.
So unless there is a valid use case for having characters being spoken outside of edit fields,
Why don't you just change the current setting when enabled to be valid only for edit fields?
In this case we could let it be a checkbox and don't need three options.

amirmahdifard commented:
@Adriani90 hi. No, I also don't agree for removal of that in my opinion because, input help is a thing, but this is stil needed for some people where they need to press some keys and need to be announced for them what they've preased and what came up, or what happened as a result, while some newbie people need to turn on input help, press a fue keys that they need, and turn of input help, and make sure they've pressed or pressing the correct key. Also, for a groupe of users with old habit, better to not remove it but rename it instead for example, from on, to everywhere. so the options are as folows. off, in edit boxes only, everywhare.

cary-rowen committed:
Merged master branch

cary-rowen commented:
@CyrilleB79 
wrote:
> the expected behaviour with the new option in ssh password case and in read-only edit fields such as synthesizer is that NVDA should not speak anything since nothing is actually visible on the screen.

I agree that there shouldn't be any feedback in the read-only edit box after enabling the new options introduced by this PR. This is currently as expected.
However, I would like to have feedback when typing the password in ssh, unlike other password fields, in the normal password field NVDA will report 'star' when typing the password.
In the SSH password field, the characters are actually entered, they are just not displayed, unlike read-only fields (which do not produce any input)

cc @seanbudd

cary-rowen commented:
I agree with @amirmahdifard I am totally not in favor of removing the original mode.

CyrilleB79 commented:
> I agree that there shouldn't be any feedback in the read-only edit box after enabling the new options introduced by this PR. This is currently as expected. However, I would like to have feedback when typing the password in ssh, unlike other password fields, in the normal password field NVDA will report 'star' when typing the password. In the SSH password field, the characters are actually entered, they are just not displayed, unlike read-only fields (which do not produce any input)

It's worth noting that visually, nothing appears on the screen. It's quite a bad design IMO, because users cannot know if and how many characters they enter (or delete) when pressing a key. Though, sighted people and blind ones are equally impacted in this case. IMO, this issue should not be a reason to hold off this PR.

amirmahdifard commented:
@cary-rowen but, if you haven't allready, please rename the "on" to everywhere" because on does not have any meaning anymore in this cace, because both ways are on, but we are only indicating where to announce the characters and words. In edit boxes only, everywhere, or off "do not announce anywhere at all". Thanks. Once that's done, this pr is completely done and ok like the addons which were tryed to impliment this feature. this is ready to go and i'm gonna excitedly wait for the next releace which I can uninstall addon and have this feature natively with in nvda it self.

cary-rowen committed:
Merge branch 'master' into typedWords

cary-rowen committed:
Merge branch 'master' into typedWords

cary-rowen committed:
Just update changes.

cary-rowen committed:
Change 'On' to 'Everywhere'

XLTechie commented:
Not very important, but
Rather than "Everywhere", I would probably call it "Always" to contrast with the "Only" of "Only in edit controls".

I personally agree with the suggestion to make "only" the new default, or upgraded version of the current "enabled" state.

By far, I think this is what users would have wanted "enabled" to mean all along, and it doesn't seem to be a very harmful change even if someone dislikes it.

XLTechie commented:
> However, I would like to have feedback when typing the password in ssh

Doesn't the "Speak passwords in all enhanced terminals" setting in Advanced, already cover this situation?

cary-rowen committed:
Merge branch 'master' into typedWords

cary-rowen commented:
@XLTechie 
I agree using "Always" which is also more common in NVDA.
> Doesn't the "Speak passwords in all enhanced terminals" setting in Advanced, already cover this situation?

Yes, I guess it makes intuitive sense that we should report passwords in this mode as long as the user explicitly enables the Report Passwords option in the advanced settings.

SaschaCowley commented on source/speech/speech.py:
```diff
@@ -780,8 +781,8 @@ def _getPlaceholderSpeechIfTextEmpty(
-		user) *and* the placeholder. The placeholder should "disappear" once the user types a value.
+	 value unless the text is empty, because it is confusing to hear the current value (presumably typed by the
+	 user) *and* the placeholder. The placeholder should "disappear" once the user types a value.
 	@return: (True, SpeechSequence) if text for obj was considered empty and we attempted to get speech for the
 		placeholder value. (False, []) if text for obj was not considered empty.
```
```suggestion
	:return: `(True, SpeechSequence)` if text for obj was considered empty and we attempted to get speech for the
		placeholder value. `(False, [])` if text for obj was not considered empty.
```

SaschaCowley commented on source/speech/speech.py:
```diff
@@ -1374,8 +1383,14 @@ def speakTypedCharacters(ch: str):
-			speakText(typedWord)
+		typingEchoMode = config.conf["keyboard"]["speakTypedWords"]
+		if typingEchoMode > TypingEcho.OFF.value and not typingIsProtected:
+			if (
+				typingEchoMode == TypingEcho.EDIT_CONTROLS.value
+				and is_editableControl()
+				or typingEchoMode == TypingEcho.EVERYWHERE.value
+			):
```
In general, you should perform the most expensive checks first when using `or`, as Python knows that if the first condition is true, it doesn't need to check the second condition. While this check shouldn't be too costly, it is run so often that this may make a measurable difference over time.
Also added an extra set of brackets for readability.
```suggestion
			if (
				typingEchoMode == TypingEcho.EVERYWHERE.value
				or (
					typingEchoMode == TypingEcho.EDIT_CONTROLS.value
					and is_editableControl()
				)
			):
```

SaschaCowley commented on source/speech/speech.py:
```diff
@@ -1387,8 +1402,15 @@ def speakTypedCharacters(ch: str):
-		speakSpelling(realChar)
+
+	typingEchoMode = config.conf["keyboard"]["speakTypedCharacters"]
+	if not suppress and typingEchoMode > TypingEcho.OFF.value and ch >= FIRST_NONCONTROL_CHAR:
```
```suggestion
	if not suppress and typingEchoMode != TypingEcho.OFF.value and ch >= FIRST_NONCONTROL_CHAR:
```

SaschaCowley commented on source/speech/speech.py:
```diff
@@ -1387,8 +1402,15 @@ def speakTypedCharacters(ch: str):
+	typingEchoMode = config.conf["keyboard"]["speakTypedCharacters"]
+	if not suppress and typingEchoMode > TypingEcho.OFF.value and ch >= FIRST_NONCONTROL_CHAR:
+		if (
+			typingEchoMode == TypingEcho.EDIT_CONTROLS.value
+			and is_editableControl()
+			or typingEchoMode == TypingEcho.EVERYWHERE.value
```
As above

SaschaCowley commented on source/speech/speech.py:
```diff
@@ -1355,6 +1356,14 @@ def _suppressSpeakTypedCharacters(number: int):
 FIRST_NONCONTROL_CHAR = " "
 
 
+def is_editableControl() -> bool:
```
Can this get a more descriptive name and a docstring?
```suggestion
def isFocusEditable() -> bool:
```

SaschaCowley commented on source/gui/settingsDialogs.py:
```diff
@@ -1974,24 +1975,27 @@ def makeSettings(self, settingsSizer):
-		self.bindHelpEvent("KeyboardSettingsSpeakTypedWords", self.wordsCheckBox)
-		self.wordsCheckBox.SetValue(config.conf["keyboard"]["speakTypedWords"])
+		# Translators: This is the label for a combobox in the keyboard settings panel.
+		charsLabelText = _("Speak typed &characters:")
```
Can the `charsX` and `wordsX` identifiers get more descriptive names? Something like `speakCharsList`.

SaschaCowley reviewed (changes requested):
Could the "everywhere" mode be renamed to "always" as others have suggested, too?

cary-rowen committed:
Apply code review changes.

cary-rowen committed:
New option is now default

CyrilleB79 commented on source/gui/settingsDialogs.py:
```diff
@@ -1974,24 +1975,27 @@ def makeSettings(self, settingsSizer):
+		self.bindHelpEvent("KeyboardSettingsSpeakTypedWords", self.speakTypedWordsList)
+		try:
+			self.speakTypedWordsList.SetSelection(config.conf["keyboard"]["speakTypedWords"])
+		except:  # noqa: E722
```
Idem

CyrilleB79 commented on source/speech/speech.py:
```diff
@@ -1374,8 +1387,16 @@ def speakTypedCharacters(ch: str):
-		if config.conf["keyboard"]["speakTypedWords"] and not typingIsProtected:
-			speakText(typedWord)
+		typingEchoMode = config.conf["keyboard"]["speakTypedWords"]
+		if typingEchoMode > TypingEcho.OFF.value and not typingIsProtected:
```
Could you be consistent with what is done later for characters, i.e. use `!=` instead of `>`? I also feel it's slightly easier to read/understand since with this syntax, you do not need to know the numeric values of each `TypingEcho` member.

```suggestion
		if typingEchoMode != TypingEcho.OFF.value and not typingIsProtected:
```

cary-rowen committed:
Apply code review changes

cary-rowen committed:
Fix name error

cary-rowen committed:
Fix lint

cary-rowen committed:
Merge branch 'master' into typedWords

pre-commit-ci[bot] committed:
Pre-commit auto-fix

cary-rowen committed:
Use 'always' instead of 'everywhere, update translator comments'

cary-rowen committed:
Uasof 'Everywhere, Update translator comments'e 'Always' insted

cary-rowen committed:
Merge branch 'TypedWords' of github.com:cary-rowen/nvda into typedWords

cary-rowen commented:
Thanks @SaschaCowley and @CyrilleB79 for the review.
For the upgrade step I'm wondering do I need to create a separate upgrade method?
For example, `upgradeConfigFrom_14_to_15`

As @CyrilleB79 commented:

> Should the removal of  		self.charsCheckBox  be documented in the API breaking changes of the change log? Idem for words?
> Though, I am not sure that we have always respected the API when modifying the GUI in non-.1 releases in the past...


I'm not sure about this.

CyrilleB79 commented on source/config/profileUpgradeSteps.py:
```diff
@@ -439,6 +440,31 @@ def upgradeConfigFrom_13_to_14(profile: ConfigObj):
 	log.debug("Deleting old config value.")
 	del profile["speech"]["outputDevice"]
 
+	# """Convert keyboard typing echo configuration from boolean to integer values."""
```
You should rather do your own profile upgrade IMO, i.e. 14 to 15. Indeed, in the following cases, the upgrade of typing echo will not be done:
* If the user is on alpha channel and has upgraded each day, i.e. has already executed profile upgrade from 13 to 14 but without the typing echo upgrade code
* overall, even upgrading from 2024.4.1 to 2025.1 (stable channel), if no output device is present in config, because the function `upgradeConfigFrom_13_to_14` will return early without evaluating typing eco code.

CyrilleB79 commented on source/config/profileUpgradeSteps.py:
```diff
@@ -439,6 +440,31 @@ def upgradeConfigFrom_13_to_14(profile: ConfigObj):
+		if configobj.validate.is_boolean(speakTypedCharacters):
+			old_value = speakTypedCharacters
+			new_value = TypingEcho.EDIT_CONTROLS.value if speakTypedCharacters else TypingEcho.OFF.value
+			profile["keyboard"]["speakTypedCharacters"] = new_value
```
A `KeyError` here (if any) should not be caught. Could you rather use a try/except/else structure, putting in the except clause only one line of code? And same for typed words upgrade below.

cary-rowen replied:
Regarding the use of the try/except/else structure:  I understand you recommend placing the code that may raise a  KeyError  within the  try  block, keeping only one line in the  except  block to handle the exception, and moving the main logic to the  else  block. Is this correct?

CyrilleB79 replied:
Yes that's correct.
Sorry my initial wording was not clear nor completely correct since there is already one line in the except block. I rather mean:
A `KeyError` here (if any) should not be caught. Could you rather use a try/except/else structure, putting in the try block only one line of code? And same for typed words upgrade below.

cary-rowen committed:
New upgrade step method

pre-commit-ci[bot] committed:
Pre-commit auto-fix

cary-rowen committed:
Merge branch 'master' into typedWords

SaschaCowley commented on source/config/configSpec.py:
```diff
@@ -178,8 +178,10 @@
+	# 0: Off, 1: Always, 2: Only in edit controls
+	speakTypedCharacters = integer(default=2,min=0,max=2)
+	# 0: Off, 1: Always, 2: Only in edit controls
+	speakTypedWords = integer(default=2,min=0,max=2)
```
Please change this back to defaulting to off (`0`).

SaschaCowley commented on source/config/configSpec.py:
Please bump the schema version to 15 so that alpha users don't get hit with a corrupt config.

cary-rowen committed:
Improve upgrade steps

SaschaCowley commented on source/globalCommands.py:
```diff
@@ -536,38 +537,32 @@ def script_previousSynthSetting(self, gesture):
-		# Translators: Input help mode message for toggle speaked typed characters command.
-		description=_("Toggles on and off the speaking of typed characters"),
+		# Translators: Input help mode message for cycling the reporting of typed words.
+		description=_("Cycle the reporting of typed characters: Off, Always, and Only in edit controls."),
```
```suggestion
		description=_("Cycles through options for when to speak typed characters."),
```

SaschaCowley commented on source/globalCommands.py:
```diff
@@ -536,38 +537,32 @@ def script_previousSynthSetting(self, gesture):
 	@script(
 		# Translators: Input help mode message for toggle speak typed words command.
-		description=_("Toggles on and off the speaking of typed words"),
+		description=_("Cycle the reporting of typed words: Off, Always, and Only in edit controls."),
```
```suggestion
		description=_("Cycles through options for when to speak typed words."),
```

SaschaCowley commented on source/gui/settingsDialogs.py:
```diff
@@ -1974,24 +1975,27 @@ def makeSettings(self, settingsSizer):
-		# Translators: This is the label for a checkbox in the
-		# keyboard settings panel.
-		charsText = _("Speak typed &characters")
-		self.charsCheckBox = sHelper.addItem(wx.CheckBox(self, label=charsText))
```
I think we should do this to be safe.

SaschaCowley commented on source/gui/settingsDialogs.py:
```diff
@@ -1974,24 +1975,29 @@ def makeSettings(self, settingsSizer):
-		# Translators: This is the label for a checkbox in the
-		# keyboard settings panel.
-		speakTypedWordsText = _("Speak typed &words")
-		self.wordsCheckBox = sHelper.addItem(wx.CheckBox(self, label=speakTypedWordsText))
```
Please also document the removal of this checkbox in the changes for developers.

SaschaCowley commented on source/speech/speech.py:
```diff
@@ -1355,6 +1356,18 @@ def _suppressSpeakTypedCharacters(number: int):
 
 
+def isFocusEditable() -> bool:
+	"""Check if the currently focused object is editable.
+	@return: C{True} if the focused object is editable, C{False} otherwise.
+	@rtype: bool
+	"""
```
Please rewrite this to be in sphinx style
```suggestion
	"""Check if the currently focused object is editable.
	:return: ``True`` if the focused object is editable, ``False`` otherwise.
	"""
```

SaschaCowley commented on user_docs/en/changes.md:
```diff
@@ -50,6 +50,7 @@ To use this feature, "allow NVDA to control the volume of other applications" mu
   * Updated CLDR to version 46.0. (#17484, @OzancanKaratas)
 * Short versions of the most commonly used command line options have been added: `-d` for `--disable-addons` and `-n` for `--lang`.
 Prefix matching on command line flags, e.g. using `--di` for `--disable-addons` is no longer supported. (#11644, @CyrilleB79)
+* The keyboard settings for "Speak typed characters" and "Speak typed words" now have three options: Off, Always, and Only in edit controls. (#17505, @Cary-rowen)
```
Can you please document that the new default is "Only in edit controls"?

SaschaCowley commented on user_docs/en/changes.md:
Please document the change of config spec, and the removal of the speak typed characters/words checkboxes from the GUI, in the changes for developers.

SaschaCowley commented on user_docs/en/userGuide.md:
```diff
@@ -2590,15 +2590,25 @@ If no key is chosen as the NVDA key it may be impossible to access many NVDA com
 
 Key: NVDA+2
```
```suggestion
Key: `NVDA+2`
```

SaschaCowley commented on user_docs/en/userGuide.md:
```diff
@@ -2590,15 +2590,25 @@ If no key is chosen as the NVDA key it may be impossible to access many NVDA com
 
 ##### Speak Typed Words {#KeyboardSettingsSpeakTypedWords}
 
 Key: NVDA+3
```
```suggestion
Key: `NVDA+3`
```

SaschaCowley reviewed (changes requested):
Please also update the copyright headers of any files you have touched to reflect that they have been modified in 2025.

cary-rowen commented on source/config/configFlags.py:
```diff
@@ -45,6 +45,30 @@ def _displayStringLabels(self):
+	"""
+
+	OFF = 0
+	ALWAYS = 1
+	EDIT_CONTROLS = 2
```
Makes sense

cary-rowen committed:
Applied some review suggestions and improved the upgrade steps

cary-rowen committed:
Update changes entries.

cary-rowen committed:
Update copyright headers

gexgd0419 committed:
Make SAPI5 & MSSP voices use WavePlayer (WASAPI) (#17592)

Closes #13284

Summary of the issue:
Currently, SAPI5 and MSSP voices use their own audio output mechanisms, instead of using the WavePlayer (WASAPI) inside NVDA.

This may make them less responsive compared to eSpeak and OneCore voices, which are using the WavePlayer, or compared to other screen readers using SAPI5 voices, according to my test result.

This also gives NVDA less control of audio output. For example, audio ducking logic inside WavePlayer cannot be applied to SAPI5 voices, so additional code is required to compensate for this.

Description of user facing changes
SAPI5 and MSSP voices will be changed to use the WavePlayer, which may make them more responsive (have less delay).

According to my test result, this can reduce the delay by at least 50ms.

This haven't trimmed the leading silence yet. If we do that also, we can expect the delay to be even less.

Description of development approach
Instead of setting self.tts.audioOutput to a real output device, do the following:

create an implementation class SynthDriverAudioStream to implement COM interface IStream, which can be used to stream in audio data from the voices.
Use an SpCustomStream object to wrap SynthDriverAudioStream and provide the wave format.
Assign the SpCustomStream object to self.tts.AudioOutputStream, so SAPI will output audio to this stream instead.
Each time an audio chunk needs to be streamed in, ISequentialStream_RemoteWrite will be called, and we just feed the audio to the player. IStream_RemoteSeek can also be called when SAPI wants to know the current byte position of the stream (dlibMove should be zero and dwOrigin should be STREAM_SEEK_CUR in this case), but it is not used to actually "seek" to a new position. IStream_Commit can be called by MSSP voices to "flush" the audio data, where we do nothing. Other methods are left unimplemented, as they are not used when acting as an audio output stream.

Previously, comtypes.client.GetEvents was used to get the event notifications. But those notifications will be routed to the main thread via the main message loop. According to the documentation of ISpNotifySource:

Note that both variations of callbacks as well as the window message notification require a window message pump to run on the thread that initialized the notification source. Callback will only be called as the result of window message processing, and will always be called on the same thread that initialized the notify source. However, using Win32 events for SAPI event notification does not require a window message pump.

Because the audio data is generated and sent via IStream on a dedicated thread, receiving events on the main thread can make synchronizing events and audio difficult.

So here SapiSink is changed to become an implementation of ISpNotifySink. Notifications received via ISpNotifySink are "free-threaded", sent on the original thread instead of being routed to the main thread.

To connect the sink, use ISpNotifySource::SetNotifySink.
To get the actual event that triggers the notification, use ISpEventSource::GetEvents. Events can contain pointers to objects or memory, so they need to be freed manually.
Finally, all audio ducking related code are removed. Now WavePlayer should be able to handle audio ducking when using SAPI5 and MSSP voices.

SaschaCowley commented on user_docs/en/changes.md:
```diff
@@ -137,6 +138,9 @@ Add-ons will need to be re-tested and have their manifest updated.
   * Added the `matchFunc` parameter to `addUsbDevices` which is also available on `addUsbDevice`.
     * This way device detection can be constrained further in cases where a VID/PID-combination is shared by multiple devices across multiple drivers, or when a HID device offers multiple endpoints, for example.
     * See the method documentation as well as examples in the albatross and brailliantB drivers for more information.
+* Changed keyboard typing echo configuration from boolean to integer values. (#17505, @Cary-rowen)
```
This is an API breaking change

SaschaCowley reviewed (changes requested):
This is looking good, just a few small things to go

cary-rowen committed:
Apply code review suggestions.

cary-rowen commented:
Thanks @SaschaCowley 
learned a lot

wmhn1872265132 commented on user_docs/en/changes.md:
```diff
@@ -51,6 +51,8 @@ To use this feature, "allow NVDA to control the volume of other applications" mu
 * Short versions of the most commonly used command line options have been added: `-d` for `--disable-addons` and `-n` for `--lang`.
 Prefix matching on command line flags, e.g. using `--di` for `--disable-addons` is no longer supported. (#11644, @CyrilleB79)
 * Microsoft Speech API version 5 and Microsoft Speech Platform voices now use WASAPI for audio output, which may improve the responsiveness of those voices. (#13284, @gexgd0419)
+* The keyboard settings for "Speak typed characters" and "Speak typed words" now have three options: Off, Always, and Only in edit controls. (#17505, @Cary-rowen)
```
The order here·Does it need to be adjusted?
```suggestion
* The keyboard settings for "Speak typed characters" and "Speak typed words" now have three options: Off, Only in edit controls, and Always. (#17505, @Cary-rowen)
```

wmhn1872265132 commented on user_docs/en/userGuide.md:
```diff
@@ -2588,17 +2588,27 @@ If no key is chosen as the NVDA key it may be impossible to access many NVDA com
+
+* Off: NVDA will not announce typed characters.
+* Always: NVDA will announce all typed characters.
+* Only in edit controls: NVDA will only announce characters typed in edit controls and other areas where text can be typed.
```
The order here·Does it need to be adjusted?

wmhn1872265132 commented on user_docs/en/userGuide.md:
```diff
@@ -2588,17 +2588,27 @@ If no key is chosen as the NVDA key it may be impossible to access many NVDA com
-When enabled, NVDA will announce all words you type on the keyboard.
+* Off: NVDA will not announce typed words.
+* Always: NVDA will announce all typed words.
+* Only in edit controls: NVDA will only announce words typed in edit controls and other areas where text can be typed.
```
The order here·Does it need to be adjusted?

cary-rowen replied:
Yes, Done

cary-rowen committed:
Adjusted the order of  options in the doc.

cary-rowen committed:
Add toggleIntegerValue helper function and update toggle scripts

pre-commit-ci[bot] committed:
Pre-commit auto-fix

cary-rowen committed:
Fix conflicts

nvdaes commented on source/globalCommands.py:
```diff
@@ -536,38 +563,36 @@ def script_previousSynthSetting(self, gesture):
 	@script(
-		# Translators: Input help mode message for toggle speaked typed characters command.
-		description=_("Toggles on and off the speaking of typed characters"),
+		# Translators: Input help mode message for cycling the reporting of typed words.
```
```suggestion
		# Translators: Input help mode message for cycling the reporting of typed characters.
```

nvdaes commented on source/globalCommands.py:
```diff
@@ -536,38 +563,36 @@ def script_previousSynthSetting(self, gesture):
+		)
 
 	@script(
 		# Translators: Input help mode message for toggle speak typed words command.
```
```suggestion
# Translators: Input help mode message for cycling the reporting of typed words.
```

nvdaes reviewed (changes requested):
@cary-rowen  I've reviewed changes in globalCommands.py.
Looks good to me, except for minor inconsistencies in comments for translators.

cary-rowen committed:
Apply code review suggestions, add type hints

pre-commit-ci[bot] committed:
Pre-commit auto-fix

cary-rowen committed:
Merge branch 'master' into TypedWords

SaschaCowley commented on source/globalCommands.py:
```diff
@@ -536,38 +563,36 @@ def script_previousSynthSetting(self, gesture):
+			enumClass=TypingEcho,
+			# Translators: Reported when the user cycles through speak typed characters modes.
+			# {mode} will be replaced with the mode; e.g. Off, On, Only in edit controls.
+			messageTemplate=_("Speak typed characters: {mode}"),
```
```suggestion
			messageTemplate=_("Speak typed characters {mode}"),
```

SaschaCowley commented on source/globalCommands.py:
```diff
@@ -536,38 +563,36 @@ def script_previousSynthSetting(self, gesture):
+			enumClass=TypingEcho,
+			# Translators: Reported when the user cycles through speak typed words modes.
+			# {mode} will be replaced with the mode; e.g. Off, On, Only in edit controls.
+			messageTemplate=_("Speak typed words: {mode}"),
```
```suggestion
			messageTemplate=_("Speak typed words {mode}"),
```

SaschaCowley committed:
Apply suggestions from code review

SaschaCowley reviewed (approved):
Thanks @cary-rowen , looks good!

cary-rowen committed:
Merge branch 'master' into TypedWords

cary-rowen commented:
I've updated the original description with the latest changes. If there is anything else I need to do before merging please let me know.
cc @SaschaCowley @seanbudd

SaschaCowley commented:
Thanks @cary-rowen, we're just waiting on @Qchristensen to look over the documentation changes.

Qchristensen reviewed (approved):
Looks good, thanks.

SaschaCowley commented:
@cary-rowen please resolve merge conflicts

cary-rowen committed:
Merge branch 'master' into TypedWords

cary-rowen committed:
Merge branch 'master' into TypedWords
